cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/gvn/internal/ValueNumberingInternal.qll

```diff
@@ -19,29 +19,26 @@ newtype TValueNumber =
     fieldAddressValueNumber(_, irFunc, field, objectAddress)
   } or
   TBinaryValueNumber(
-    IRFunction irFunc, Opcode opcode, IRType type, TValueNumber leftOperand,
-    TValueNumber rightOperand
+    IRFunction irFunc, Opcode opcode, TValueNumber leftOperand, TValueNumber rightOperand
   ) {
-    binaryValueNumber(_, irFunc, opcode, type, leftOperand, rightOperand)
+    binaryValueNumber(_, irFunc, opcode, leftOperand, rightOperand)
   } or
   TPointerArithmeticValueNumber(
-    IRFunction irFunc, Opcode opcode, IRType type, int elementSize, TValueNumber leftOperand,
+    IRFunction irFunc, Opcode opcode, int elementSize, TValueNumber leftOperand,
     TValueNumber rightOperand
   ) {
-    pointerArithmeticValueNumber(_, irFunc, opcode, type, elementSize, leftOperand, rightOperand)
+    pointerArithmeticValueNumber(_, irFunc, opcode, elementSize, leftOperand, rightOperand)
   } or
-  TUnaryValueNumber(IRFunction irFunc, Opcode opcode, IRType type, TValueNumber operand) {
-    unaryValueNumber(_, irFunc, opcode, type, operand)
+  TUnaryValueNumber(IRFunction irFunc, Opcode opcode, TValueNumber operand) {
+    unaryValueNumber(_, irFunc, opcode, operand)
   } or
   TInheritanceConversionValueNumber(
     IRFunction irFunc, Opcode opcode, Class baseClass, Class derivedClass, TValueNumber operand
   ) {
     inheritanceConversionValueNumber(_, irFunc, opcode, baseClass, derivedClass, operand)
   } or
-  TLoadTotalOverlapValueNumber(
-    IRFunction irFunc, IRType type, TValueNumber memOperand, TValueNumber operand
-  ) {
-    loadTotalOverlapValueNumber(_, irFunc, type, memOperand, operand)
+  TLoadTotalOverlapValueNumber(IRFunction irFunc, TValueNumber memOperand, TValueNumber operand) {
+    loadTotalOverlapValueNumber(_, irFunc, memOperand, operand)
```

Comment:
Will the removal of `type` here make us confused about the following (inspired by https://github.com/Semmle/ql/pull/2772#discussion_r375498142)?
```
double foo(int* p, int choice) {
  if (choice == 1)
    return *(int*)p; // 1
  else if (choice == 2)
    return (int)*(char*)p; // 2
  else
    return *(float*)p; // 3
}
```

I think one could argue that 1 and 3 should have the same value number before their conversion to `double` since they're the same bit pattern, but it seems wrong to give 1 and 2 the same value number. It seems especially wrong to give the same value number to the `*` in 1 and the `(int)` in 2.

Reply:
Adding the following test function to `ir_gvn.ql`:

void foo(int* p) {
  double d1 = *(int*)p;
  double d2 = (int)*(char*)p;
  double d3 = *(float*)p;
}

gives the following value numbers with this PR:

```
r162_5 = *(int*)p // r162_5, r163_5, r164_5
r163_5 = *(char*)p // r162_5, r163_5, r164_5
r162_6 = (double)*(int*)p  // r162_6, r163_6, r164_6
r163_6 = (int)*(char*)p // r162_6, r163_6, r164_6
r164_5 = *(float*)p // r162_5, r163_5, r164_5
r164_6 = (double)*(float*)p // r162_6, r163_6, r164_6
```

So the expressions `*(int*)p` and `*(float*)p` have the same value number. (i.e., `// 1` and `// 3` have the same value number before conversions to `double`).
But `d1` and `d2` are assigned the same value number, sadly.

Reply:
At least based on the current design, `*(int*)p` and `*(float*)p` should _not_ have the same value number. The intention is that two results will only have the same value number if they have the same bit pattern and have the same `IRType`.

Reply:
Can someone check if the removal of `type` here is necessary for performance of bloomberg/bde?

Reply:
The removal is unnecessary for performance on bloomberg/bde, but we also need #2844 for it.